openeo-geotrellis/src/main/scala/org/openeo/geotrellis/package.scala

```diff
@@ -158,28 +158,63 @@ package object geotrellis {
-    @tailrec
-    def attempt(retries: Int, delay: (Long, TimeUnit))(action: => scalaj.http.HttpResponse[R]): scalaj.http.HttpResponse[R] = {
-      val (amount, timeUnit) = delay
-
-      val ret = action
-      if (ret.code == 429) {
-        logger.warn(f"Got ${ret.code}. Will retry request: " + ret)
-        // TODO: Could use Retry-After header.
-        // It is not returned by https://services.terrascope.be/ at the moment.
-        // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
-      } else {
-        return ret
-      }
-      timeUnit.sleep(amount)
-      attempt(retries - 1, (amount * 2, timeUnit)) {
-        action
-      }
-    }
 
-    attempt(retries = 4, delay = (30, SECONDS)) {
-      action
+  /**
+   * taken from DefaultProcessApi
+   * Specific to retry HttpResponse, without throwError logic.
+   */
+  def withRetryAfterRetries[R](context: String)(httpResponseCallback: => HttpResponse[R])(implicit logger: Logger): HttpResponse[R] = {
+    val retryableResult: HttpResponse[R] => Boolean = r => r.is5xx
+    val retryableException: Throwable => Boolean = {
+      case _: HttpStatusException =>
+        logger.warn("withRetryAfterRetries does not support .throwError logic.")
+        false
+      case _: SocketTimeoutException => true
+      case _: SocketException => true // Got this when terminating test server in the middle of a request
+      case e => logger.error(s"Not attempting to retry unrecoverable error in context: '$context' " + e.getMessage); false
     }
+
+    val shakyConnectionRetryPolicy = new FailsafeRetryPolicy[HttpResponse[R]]()
+      .handleResultIf(retryableResult.asJava)
+      .handleIf(retryableException.asJava)
+      .withBackoff(1, 1000, ChronoUnit.SECONDS) // should not reach maxDelay because of maxAttempts 5
+      .withJitter(0.5)
+      .withMaxAttempts(5)
+      .onFailedAttempt((attempt: ExecutionAttemptedEvent[HttpResponse[R]]) => {
+        val msg = if (attempt.getLastFailure != null) {
+          attempt.getLastFailure
+        } else if (attempt.getLastResult != null) {
+          "result code: " + attempt.getLastResult.code
+        } else ""
+        logger.warn(s"Attempt ${attempt.getAttemptCount} failed in context: '$context' " + msg)
+      })
+      .onFailure((execution: ExecutionCompletedEvent[HttpResponse[R]]) => {
+        val e = execution.getFailure
+        logger.error(s"Failed after ${execution.getAttemptCount} attempt(s) in context: '$context'" + e.getMessage)
+      })
+
+    val isRateLimitingResponse: HttpResponse[R] => Boolean = r => r.code == 429
+    val isRateLimitingException: Throwable => Boolean = _ => false // handle errors in shakyConnectionRetryPolicy
+    val rateLimitingRetryPolicy = new FailsafeRetryPolicy[HttpResponse[R]]()
+      .handleResultIf(isRateLimitingResponse.asJava)
+      .handleIf(isRateLimitingException.asJava)
+      .withMaxAttempts(5)
```

Comment:
```suggestion
    val isRateLimitingResponse: (HttpResponse[R], Throwable) => Boolean =
      (response, _ /* ignore exceptions, those are handled in shakyConnectionRetryPolicy */) => response.code == 429

    val rateLimitingRetryPolicy = new FailsafeRetryPolicy[HttpResponse[R]]()
      .handleIf(isRateLimitingResponse.asJava)
```

Reply:
Untested but I think this more clearly conveys that you're suppressing the default behavior of retrying on exceptions.

Reply:
That works indeed. Got trough in normal case, and fails clearly when .throwException is used in callback